src/credentials/GenerateChipX509Cert.cpp

```diff
@@ -0,0 +1,479 @@
+        ASN1_START_CONSTRUCTED(kASN1TagClass_ContextSpecific, 0)
+        {
+            // Version ::= INTEGER { v1(0), v2(1), v3(2) }
+            ASN1_ENCODE_INTEGER(2);
+        }
+        ASN1_END_CONSTRUCTED;
+
+        ReturnErrorOnFailure(writer.PutInteger(requestParams.SerialNumber));
+
+        ASN1_START_SEQUENCE { ASN1_ENCODE_OBJECT_ID(kOID_SigAlgo_ECDSAWithSHA256); }
+        ASN1_END_SEQUENCE;
+
+        // Issuer OID depends on if cert is being signed by the root CA
+        if (issuerLevel == kIssuerIsRootCA)
+        {
+            dnParams[0].AttrOID = chip::ASN1::kOID_AttributeType_ChipRootId;
+        }
+        else
+        {
+            dnParams[0].AttrOID = chip::ASN1::kOID_AttributeType_ChipICAId;
+        }
+        dnParams[0].Value = requestParams.Issuer;
+
+        if (requestParams.HasFabricID)
+        {
+            dnParams[1].AttrOID = chip::ASN1::kOID_AttributeType_ChipFabricId;
+            dnParams[1].Value   = requestParams.FabricID;
+            numDNs              = 2;
+        }
+        ReturnErrorOnFailure(EncodeChipDNs(dnParams, numDNs, writer));
+
+        // validity Validity,
+        ReturnErrorOnFailure(EncodeValidity(requestParams.ValidityStart, requestParams.ValidityEnd, writer));
+
+        // subject Name
+        if (requestParams.HasNodeID)
+        {
+            dnParams[0].AttrOID = chip::ASN1::kOID_AttributeType_ChipNodeId;
+
+            isCA = false;
+        }
+        else if (subjectPubkey != issuerPubkey)
+        {
+            dnParams[0].AttrOID = chip::ASN1::kOID_AttributeType_ChipICAId;
+        }
+        dnParams[0].Value = subject;
+
+        ReturnErrorOnFailure(EncodeChipDNs(dnParams, numDNs, writer));
+
+        ReturnErrorOnFailure(EncodeSubjectPublicKeyInfo(subjectPubkey, writer));
+
+        // certificate extensions
+        ReturnErrorOnFailure(EncodeExtensions(isCA, subjectPubkey, issuerPubkey, writer));
+    }
+    ASN1_END_SEQUENCE;
+
+exit:
+    return err;
+}
+
+static CHIP_ERROR NewChipX509Cert(const X509CertRequestParams & requestParams, CertificateIssuerLevel issuerLevel, uint64_t subject,
+                                  const Crypto::P256PublicKey & subjectPubkey, Crypto::P256Keypair & issuerKeypair,
+                                  uint8_t * x509CertBuf, uint32_t x509CertBufSize, uint32_t & x509CertLen)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    ASN1Writer writer;
+    writer.Init(x509CertBuf, x509CertBufSize);
+
+    ReturnErrorOnFailure(EncodeTBSCert(requestParams, issuerLevel, subject, subjectPubkey, issuerKeypair.Pubkey(), writer));
+    writer.Finalize();
+
+    Crypto::P256ECDSASignature signature;
+    ReturnErrorOnFailure(issuerKeypair.ECDSA_sign_msg(x509CertBuf, writer.GetLengthWritten(), signature));
+
+    writer.Init(x509CertBuf, x509CertBufSize);
+
+    ASN1_START_SEQUENCE
+    {
+        ReturnErrorOnFailure(EncodeTBSCert(requestParams, issuerLevel, subject, subjectPubkey, issuerKeypair.Pubkey(), writer));
+        SuccessOrExit(err);
```

Comment:
not needed

Reply:
This is needed, as we first generate the TBS by itself, compute the signature. The TBS is generated again inside the  `ASN1_START_SEQUENCE` along with the rest of the certificate.

Maybe we can`memmove` the TBS (generated before signing) into the final ASN1 structure. It just gets complicated since we are using the same buffer for TBS (for signing) and final ASN1 cert structure.